lib/server.js

```diff
@@ -0,0 +1,187 @@
+    var buf = message; //really necessary when we're not using arraybuffer?
+    debug(buf.compare(message));
+    debug(buf);
+    switch (message[0]) {
+      case msgcode.pause:
+        this.strbuf.pause();
+        debug('paused');
+        break;
+
+      case msgcode.resume:
+        this.strbuf.resume();
+        debug('resumed');
+        break;
+
+      case msgcode.stdin:
+        if (!this.execStream.write(message.slice(1), {binary: true})) {
+          this.sendMessage(msgcode.pause, new Buffer(0));
+          debug('paused');
+        }
+        // debug('server needs drain: ',!tf);
+        break;
+
+      default:
+        debug('unknown msg code %s', message[0]);
+    }
+  }
+}
+
+// Makes a new buffer with code number (ideally 0x00-0xFF)
+// appended to the start of the given buffer
+
+/** Makes an instance of docker exec running command, then links the provided
+    websocket to the stream provided by docker exec
+
+    @param tty: boolean representing whether or not terminal emulation is active
+  */
+
+export default class DockerExecWebsocketServer {
+  /* Creates Docker Exec instance on given container, running the first message given
+   * as a command.
+   * Options:
+   * port, required
+   * containerId, name or id of docker container, required
+   * path, path where the websocket is hosted
+   * dockerSocket, path to docker's remote API
+   */
+  constructor (options) {
+    options = _.defaults({}, options, {path: '/'+slugid.v4(),
+      dockerSocket: '/var/run/docker.sock',
+    });
+
+    //setting up docker
+    var stats = fs.statSync(options.dockerSocket);
+    if (!stats.isSocket()) {
+      throw new Error('Are you sure the docker is running?');
+    }
+    var docker = new Docker({socketPath: options.dockerSocket});
+
+    //getting container
+    assert(options.containerId, 'required container option missing');
+    var container = docker.getContainer(options.containerId);
+    assert(container, 'could not get container from Docker');
+
+    this.port = options.port;
+    this.path = options.path;
+
+    //making websocket server
+    var wsopts = {
+      port: this.port,
+      path: this.path,
+    };
+    assert(this.port, 'required port option missing');
+    this.server = new ws.Server(wsopts);
+    debug('%s%s created', wsopts.port, wsopts.path);
+
+    this.sessions = [];
+
+    this.server.on('connection', (socket) => {
+      debug('connection recieved');
+      var args = url.parse(socket.upgradeReq.url, true).query;
+      var session = new ExecSession({
+        container: container,
+        socket: socket,
+        command: args.command,
+        tty: /^true$/i.test(args.tty),
+      });
+      this.sessions.push(session);
+      session.execute();
+    });
+  }
+
+  close () {
+    this.server.close();
+    this.sessions.foreach((session) => { //could be improved later
+      session.close();
```

Comment:
You probably need to implement `session.close()`
I would make it `async` so you can do things like interrupt or try to close websockets nicely.

then where you just do `await Promise.all(this.sessions.map(s => s.close()));` of course then this `close` method has to be async too...